external-crates/move/crates/move-compiler/src/typing/translate.rs

```diff
@@ -2104,213 +2071,651 @@ fn add_field_types<T>(
+            if index_opt.is_none() {
+                context
+                    .env
+                    .add_diag(diag!(Declarations::MissingSyntaxMethod, (loc, msg()),));
+            }
+            index_opt
+        }
+        sp!(_, T::Apply(_, sp!(_, TN::Builtin(bt_name)), _)) => {
+            let index_opt = core::find_builtin_index_funs(context, loc, bt_name);
+            if index_opt.is_none() {
+                context
+                    .env
+                    .add_diag(diag!(Declarations::MissingSyntaxMethod, (loc, msg()),));
+            }
+            index_opt
+        }
+        t => {
+            let smsg = format!(
+                "Expected a struct or builtin type but got: {}",
+                core::error_format(t, &context.subst)
+            );
+            context.env.add_diag(diag!(
+                TypeSafety::ExpectedSpecificType,
+                (loc, msg()),
+                (t.loc, smsg),
+            ));
+            None
+        }
+    }
+}
+
+fn resolve_index_funs_and_type(
+    context: &mut Context,
+    loc: Loc,
+    ty: Type,
+    argloc: Loc,
+    args: &[T::Exp],
+) -> (Option<IndexSyntaxMethods>, Type) {
+    let ty_str = core::error_format(&ty, &context.subst);
+    let msg = || {
+        format!(
+            "No valid '{}({})' method found for {}",
+            SyntaxAttribute::SYNTAX,
+            SyntaxAttribute::INDEX,
+            ty_str
+        )
+    };
+    let readied = core::ready_tvars(&context.subst, ty);
+    let Some(index) = find_index_funs(context, loc, &readied) else {
+        return (None, context.error_type(loc));
+    };
+    let Some((m, f)) = index.get_name_for_typing() else {
+        context
+            .env
+            .add_diag(diag!(Declarations::MissingSyntaxMethod, (loc, msg()),));
+        return (None, context.error_type(loc));
+    };
+    let fty = core::make_function_type(context, loc, &m, &f, None);
+    let mut arg_types = args
+        .iter()
+        .map(|e| core::ready_tvars(&context.subst, e.ty.clone()))
+        .collect::<Vec<_>>();
+    // We insert a mut ref here because it will be a correct subtype regardless of if
+    // only `index` or `index_mut` is defined.
+    arg_types.insert(0, sp(loc, Type_::Ref(true, Box::new(readied))));
+    let ty = syntax_call_return_ty(context, loc, m, f, fty, argloc, arg_types);
+    (Some(index), ty)
+}
+
+//--------------------------------------------------------------------------------------------------
+// Dotted Expressions
+//--------------------------------------------------------------------------------------------------
+
+// An ExpDotted is a base expression with some number of field and index accessors attached:
+//
+//   E := base | E . field | E [ args ]
+
+#[derive(Debug)]
+enum ExpDottedAccess {
+    Field(Field, /* base type */ Type),
+    Index {
+        index_loc: Loc,
+        syntax_methods: Option<IndexSyntaxMethods>,
```

Comment:
We do like to clone a lot, but this feels like a bit of a big one. Possible to make this a reference of some sort?

Reply:
I cannot: the borrow would be from the ProgramInfo, which the Context holds. As it is, I would need to borrow the entire context, which is a non-starter due to the code structure. If I could somehow just borrow the ProgramInfo, that also would not work because you couldn't borrow it twice for subsequent index expressions on the same dotted expression.

As for size, I agree this feels heavy. That said, it _is_ only holding onto about two use funs worth of size. Luckily, these are short-lived: we process the dotted expression into this form and immediately discharge it in every case.